homeassistant/components/madvr/config_flow.py

```diff
@@ -36,81 +32,113 @@ class MadVRConfigFlow(ConfigFlow, domain=DOMAIN):
+        return await self._handle_config_step(user_input)
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle reconfiguration of the device."""
+        self.entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
+        return await self.async_step_reconfigure_confirm(user_input)
+
+    async def async_step_reconfigure_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a reconfiguration flow initialized by the user."""
+        return await self._handle_config_step(user_input, step_id="reconfigure")
+
+    async def _handle_config_step(
+        self, user_input: dict[str, Any] | None = None, step_id: str = "user"
+    ) -> ConfigFlowResult:
+        """Handle the configuration step for both initial setup and reconfiguration."""
         errors: dict[str, str] = {}
 
         if user_input is not None:
+            _LOGGER.debug("User input: %s", user_input)
             host = user_input[CONF_HOST]
             port = user_input[CONF_PORT]
 
             try:
-                # ensure we can connect and get the mac address from device
-                mac = await self._test_connection(host, port)
+                mac = await test_connection(self.hass, host, port)
             except CannotConnect:
                 _LOGGER.error("CannotConnect error caught")
                 errors["base"] = "cannot_connect"
             else:
                 if not mac:
                     errors["base"] = "no_mac"
-            if not errors:
-                _LOGGER.debug("MAC address found: %s", mac)
-                # this will prevent the user from adding the same device twice and persist the mac address
-                await self.async_set_unique_id(mac)
-                self._abort_if_unique_id_configured()
-
-                # create the entry
-                return self.async_create_entry(
-                    title=DEFAULT_NAME,
-                    data=user_input,
-                )
-
-        # this will show the form or allow the user to retry if there was an error
+                else:
+                    _LOGGER.debug("MAC address found: %s", mac)
+                    # if mac does change, it should be updated
+                    await self.async_set_unique_id(mac)
+
+                    if step_id == "reconfigure" and self.entry:
+                        _LOGGER.debug("Reconfiguration done")
+                        return self.async_update_reload_and_abort(
+                            entry=self.entry,
+                            data={**user_input, CONF_HOST: host, CONF_PORT: port},
+                            reason="reconfigure_successful",
+                        )
+
+                    # however dont abort if not unique because the IP could change but keep the same mac
+                    self._abort_if_unique_id_configured()
```

Comment:
You can auto update iirc

Reply:
for what purpose? I don't want to abort if the IP changes but the mac is the same

Reply:
But you do want to update the IP address in that case and not create a second entry

Reply:
okay I think I did what you are saying but not really clear because its undocumented

if the IP changes during reconfigure, it will always update the entry

if its a new mac, it will update the unique ID

I am not sure what else _abort_if_unique_id_configured should be doing since it only matters during a new configuration